// @moneyspace.net/moneyspace-node-js-sdk/dist/tests/api.manual.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const http = require("http");
const os = require("os");
const path = require("path");
const lib_1 = require("../lib");
function run() {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        const baseUrl = process.env['HOST'] || undefined;
        console.log(`Base url: ${baseUrl}`);
        const api = lib_1.createMoneySpace({
            baseUrl,
            credentials: lib_1.createFileCredentialsProvider(path.join(os.homedir(), '.moneyspace-test-creds.json')),
            successUrl: 'https://www.moneyspace.net/merchantapi/paycardsuccess',
            cancelUrl: 'https://www.moneyspace.net?status=cancel',
            failUrl: 'https://www.moneyspace.net?status=fail'
        });
        const port = 8080;
        const webHookHandler = api.createWebHookRequestHandler();
        const server = http.createServer((req, resp) => {
            webHookHandler(req, resp);
        });
        server
            .on('error', function (e) {
            console.error(e);
        })
            .listen({
            host: '0.0.0.0',
            port
        }, () => {
            console.log(`HTTP Server running at ${port} on 0.0.0.0`);
        });
    });
}
run();
