// @moneyspace.net/moneyspace-node-js-sdk/dist/tests/api.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const ava_1 = require("ava");
const nock = require("nock");
const lib_1 = require("../lib");
const qs = require("query-string");
let nockScope;
function createApi() {
    return lib_1.createMoneySpace({
        credentials: lib_1.createSimpleCredentialsProvider('secret:e5e362183961480383d7a9662cbb3a5f', 'key:2e5475b2ea40440b94176f4c138d21d6'),
        successUrl: 'https://www.moneyspace.net/merchantapi/paycardsuccess',
        cancelUrl: 'https://www.moneyspace.net?status=cancel',
        failUrl: 'https://www.moneyspace.net?status=fail'
    });
}
ava_1.before(() => {
    nockScope = nock('https://www.moneyspace.net');
    nock.disableNetConnect();
});
ava_1.after(() => {
    nock.cleanAll();
    nock.enableNetConnect();
});
ava_1.default('mocked payment transaction', (t) => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
    nockScope.post('/merchantapi/v2/createpayment/obj').reply(200, () => {
        return [
            {
                status: 'Create Success',
                'Transaction ID': 'MSTRF18000000017437'
            }
        ];
    });
    const api = createApi();
    const resp = yield api.merchantApi.sendPaymentTransaction({
        amount: '0.15',
        customerOrderId: 'f229c529d4c347449d6fb09c51b979a0',
        firstName: 'John',
        lastName: 'Doe'
    });
    t.is(resp.orderId, 'f229c529d4c347449d6fb09c51b979a0');
    t.is(resp.transactionId, 'MSTRF18000000017437');
    t.truthy(resp.timeHash);
}));
ava_1.default('mocked payment rejection', (t) => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
    nockScope.post('/merchantapi/v2/createpayment/obj').reply(200, () => {
        return [
            {
                status: 'NotFound'
            }
        ];
    });
    const api = createApi();
    const err = yield t.throwsAsync(api.merchantApi.sendPaymentTransaction({
        amount: '0.15',
        customerOrderId: 'f229c529d4c347449d6fb09c51b979a0',
        firstName: 'John',
        lastName: 'Doe'
    }));
    t.is(err.message, 'NotFound');
}));
ava_1.default('mocked check payment transaction', (t) => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
    nockScope.post('/merchantapi/v1/findbytransaction/obj').reply(200, (uri, requestBody) => {
        const body = qs.parse(requestBody.toString());
        return [
            {
                Amount: '100.25',
                'transaction ID': body.transactionID,
                'Status Payment': 'pending',
                Description: 'Tx description'
            }
        ];
    });
    const api = createApi();
    const resp = yield api.merchantApi.checkPaymentTransaction({
        transactionId: 'MSTRF18000000017437'
    });
    t.is(resp.transactionId, 'MSTRF18000000017437');
    t.is(resp.amount, '100.25');
}));
ava_1.default('mocked check payment transaction rejection', (t) => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
    nockScope.post('/merchantapi/v1/findbytransaction/obj').reply(200, (uri, requestBody) => {
        return [
            {
                status: 'You cannot have permission to view this transaction.'
            }
        ];
    });
    const api = createApi();
    const err = yield t.throwsAsync(api.merchantApi.checkPaymentTransaction({
        transactionId: 'MSTRF18000000017437'
    }));
    t.is(err.message, 'You cannot have permission to view this transaction.');
}));
ava_1.default('mocked check order id', (t) => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
    nockScope.post('/merchantapi/v1/findbyorderid/obj').reply(200, (uri, requestBody) => {
        const body = qs.parse(requestBody.toString());
        return [
            {
                Amount: '100.25',
                'Order ID': body.orderID,
                'Status Payment': 'pending',
                Description: 'Tx description'
            }
        ];
    });
    const api = createApi();
    const resp = yield api.merchantApi.checkPaymentTransaction({
        orderId: 'd8a3287c'
    });
    t.is(resp.orderId, 'd8a3287c');
    t.is(resp.amount, '100.25');
}));
ava_1.default('mocked check merchant profile', (t) => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
    nockScope.get(/\/merchantapi\/v1\/store\/obj.*/).reply(200, () => {
        return [
            {
                Store: [
                    {
                        name: 'test store',
                        logo: 'https://moneyspace.s3-ap-southeast-1.amazonaws.com/1194_20190707220401917.jpeg',
                        telephone: '0999999903'
                    }
                ]
            }
        ];
    });
    const api = createApi();
    const resp = yield api.merchantApi.getMerchantProfile();
    t.true(resp.stores && Array.isArray(resp.stores));
    const store = resp.stores[0] || {};
    t.is(store.name, 'test store');
    t.is(store.logo, 'https://moneyspace.s3-ap-southeast-1.amazonaws.com/1194_20190707220401917.jpeg');
    t.is(store.phone, '0999999903');
}));
